### PR TITLE
Tests + Print services

### PR DESCRIPTION
**Pull request by procrastinatio**
Adding some test to the print service

[Demo](https://mf-geoadmin3.int.bgdi.ch/mom_print_service/index.html)

### CODE REVIEW

**Comment by loicgasser** [comment]
much nicer

**Comment by gjn** [comment]
Could you squash it?

File: src/components/print/PrintLayerService.js [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+            encLayer = encodeVector(layer,
+                src.getFeatures(),
+                scaleDenom,
+                printRectangeCoords,
+                dpi);
+          }
+        }
+      } else {
+          console.error('Trying to encode a group with the layer encoder!');
+
```
Comment (loicgasser) [inline src/components/print/PrintLayerService.js]: blank line

File: test/specs/print/PrintLayerService.spec.js [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+    describe('#encodeDimensions()', function() {
+
+      var params = {
+        Time: "current"
+      }
+
+      it('returns an encoded Dimensions', function() {
+
+
```
Comment (loicgasser) [inline test/specs/print/PrintLayerService.spec.js]: blank

File: test/specs/print/PrintLayerService.spec.js [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+          "dimensions": ["TIME"],
+          "params": {
+            "TIME": "current"
+          },
+          "matrixSet": "21781",
+          "baseURL": "https://wmts.geo.admin.ch",
+          "zoomOffset": 0,
+          "tileOrigin": [420000, 350000],
+          "tileSize": [256, 256],
+          "resolutions": [4000, 3750, 3500, 3250, 3000, 2750, 2500, 2250, 2000, 1750, 1500, 1250, 1000, 750, 650, 500, 250, 100, 50, 20, 10, 5, 2.5, 2, 1.5],
```
Comment (loicgasser) [inline test/specs/print/PrintLayerService.spec.js]: you could define these variables once at the top of this file. They are repeated over and over again.

File: test/specs/print/PrintLayerService.spec.js [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        expect(encLegend.classes[0].icon).to.eql(
+           options.legendUrl +  'ch.swisstopo.swisstlm3d-wanderwege_fr.png'
+        );
+        stub.restore();
+      });
+    });
+    
+    describe('#encodeDimensions()', function() {
+        
+        var geometry = new ol.geom.Point(
```
Comment (loicgasser) [inline test/specs/print/PrintLayerService.spec.js]: indent

File: test/specs/print/PrintLayerService.spec.js [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+
+      var options = {
+        url: "https://wms.geo.admin.ch/?",
+        projection: "EPSG:21781",
+        ratio: 1,
+        params: {
+          LAYERS: "ch.swisstopo.fixpunkte-agnes"
+        }
+
+
```
Comment (loicgasser) [inline test/specs/print/PrintLayerService.spec.js]: blank

File: src/components/print/PrintLayerService.js [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+(function() {
+
+  var module = angular.module('ga_printlayer_service', [
+    'ga_translation_service',
+    'ga_urlutils_service',
+    'ga_printstyle_service',
+    'pascalprecht.translate'
+  ])
+      .provider('gaPrintLayer', gaPrintLayer);
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: Please use same syntax as in another services.

File: src/components/print/PrintLayerService.js [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+
+
+  var pdfLegendString = '_big.pdf';
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

File: src/components/print/PrintDirective.js [hunk 8]
```diff
@@ -694,102 +322,23 @@ goog.require('ga_urlutils_service');
...
-            },
-            'name': 'drawing',
-            'opacity': 1
-          };
-          encLayers.push(encOverlayLayer);
+        if (encOverlayLayer) {
+            encLayers.push(encOverlayLayer);
         }
-      });
+       });
```
Comment (oterral) [inline src/components/print/PrintDirective.js]: indent

File: src/components/print/PrintLayerService.js [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+
+     return {layer: encLayer, legend: encLegend};
+    }
+  };
+
+  function encodeBase(oLayer) {
+    var enc = {
+      opacity: oLayer.getOpacity()
+    };
+    if (oLayer.bodId) enc.layer = oLayer.bodId;
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: Use `{ }`

File: src/components/print/PrintLayerService.js [hunk 10]
```diff
@@ -0,0 +1,621 @@
...
+
+    return enc;
+  };
+
+
+
+  function getEncodeVector(gaPrintStyle) {
+
+      return encodeVector;
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

File: src/components/print/PrintLayerService.js [hunk 11]
```diff
@@ -0,0 +1,621 @@
...
+
+    function encodeVector(layer, features, scale, printRectangle, dpi) {
+
+      // var enc = $scope.encoders.layers['Layer'].call(this, layer);
+
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: Remove lines

File: src/components/print/PrintLayerService.js [hunk 12]
```diff
@@ -0,0 +1,621 @@
...
+    var url = (source.getUrls && source.getUrls()[0]) ||
+        (source.getUrl && source.getUrl());
+
+    var customParams = {
+        'EXCEPTIONS': 'XML',
+        'TRANSPARENT': 'true',
+        'CRS': source.getProjection().getCode(),
+        'MAP_RESOLUTION': dpi
+      };
+     if (params.TIME) customParams['TIME'] = params.TIME;
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: Use `{}`

File: src/components/print/PrintModule.js [hunk 13]
```diff
@@ -1,12 +1,14 @@
 goog.provide('ga_print');
 
 goog.require('ga_print_directive');
+goog.require('ga_printlayer_service');
 goog.require('ga_printstyle_service');
```
Comment (oterral) [inline src/components/print/PrintModule.js]: Remove the require

File: test/specs/print/PrintLayerService.spec.js [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+        $translate = $injector.get('$translate');
+        gaPrintLayer = $injector.get('gaPrintLayer');
+        gaLang = $injector.get('gaLang');
+        gaMarkerOverlay = $injector.get('gaMarkerOverlay');
+        gaUrlUtils = $injector.get('gaUrlUtils');
+      });
+    });
+
+
+    describe('#getEncodeLegend()', function() {
```
Comment (oterral) [inline test/specs/print/PrintLayerService.spec.js]: `#encodeLegend()`

File: test/specs/print/PrintLayerService.spec.js [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+          "baseURL": "https://wmts.geo.admin.ch",
+          "zoomOffset": 0,
+          "tileOrigin": [420000, 350000],
+          "tileSize": [256, 256],
+          "resolutions": [4000, 3750, 3500, 3250, 3000, 2750, 2500, 2250, 2000, 1750, 1500, 1250, 1000, 750, 650, 500, 250, 100, 50, 20, 10, 5, 2.5, 2, 1.5],
+          "maxExtent": [420000, 30000, 900000, 350000]
+        });
+      });
+    });
+    //-------------------------
```
Comment (oterral) [inline test/specs/print/PrintLayerService.spec.js]: Remove this line

File: test/specs/print/PrintLayerService.spec.js [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+    describe('#encodeWMTS()', function() {
```
Comment (oterral) [inline test/specs/print/PrintLayerService.spec.js]: You already have `#encodeWMTS()`  section  move the  `it.` to the previous section and describe what is different with the previous WMTS you encode.

File: test/specs/print/PrintLayerService.spec.js [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+      it('returns an encoded Dimensions', function() {
+
+
+        var dimensions = gaPrintLayer.encodeDimensions(params)
+
+        expect(dimensions).to.eql({ TIME: 'current' });
+      });
+    });
+
+    describe('#encodeWMSLayer()', function() {
```
Comment (oterral) [inline test/specs/print/PrintLayerService.spec.js]: `#encodeWMS()`

File: test/specs/print/PrintLayerService.spec.js [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+            "MAP_RESOLUTION": "150"
+          },
+          "singleTile": false
+        });
+      });
+    });
+
+
+
+    describe('#getMatrixIds()', function() {
```
Comment (oterral) [inline test/specs/print/PrintLayerService.spec.js]: `#encodeMatrixIds()`

File: src/components/print/PrintLayerService.js [hunk 19]
```diff
@@ -0,0 +1,621 @@
...
+   var format = new ol.format.GeoJSON();
+
+
+  function gaPrintLayer() {
+    this.$get = function($document, $translate, gaGlobalOptions,
+        gaLayers, gaTime, gaLang, gaPrintStyle, gaMapUtils,
+        gaUrlUtils) {
+      return {
+          encodeLegend: getEncodeLegend(gaLang),
+          matrixIds: getMatrixIds(gaMapUtils),
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: Why not `encodeMatrixIds` like others ?

File: src/components/print/PrintDirective.js [hunk 20]
```diff
@@ -623,13 +215,20 @@ goog.require('ga_urlutils_service');
...
+      var resolution = $scope.map.getView().getResolution();
+
+
       // Re order layer by z-index
       layers.sort(function(a, b) {
         return a.getZIndex() - b.getZIndex();
       });
 
       // Transform layers to literal
       layers.forEach(function(layer) {
```
Comment (oterral) [inline src/components/print/PrintDirective.js]: Any chance to have an `encodeLayers(layers)`  function ?  Could be in another PR.

Comment (procrastinatio) [inline src/components/print/PrintDirective.js]: Hem, it's already there, with `layer` beeing a `ol.Layer` or a `ol.LayerGroup` (recursive)
```
  encodeGroup: getEncodeGroup(gaLayers, gaPrintStyle, gaTime, gaMapUtils),
  encodeLayer: getEncodeLayer(gaLayers, gaPrintStyle, gaTime, gaMapUtils),
```

File: src/components/print/PrintLayerService.js (hunk 19 above)
Comment (procrastinatio) [inline src/components/print/PrintLayerService.js]: Artistic liberty...

File: src/components/print/PrintLayerService.js [hunk 21]
```diff
@@ -0,0 +1,594 @@
...
+  var UNITS_RATIO = 39.37; // inches per meter
+  var styleId = 0;
+  var format = new ol.format.GeoJSON();
+
+  function gaPrintLayer() {
+    this.$get = function($document, $translate, gaGlobalOptions,
+        gaLayers, gaTime, gaLang, gaPrintStyle, gaMapUtils,
+        gaUrlUtils) {
+      return {
+          encodeLegend: getEncodeLegend(gaLang),
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

File: src/components/print/PrintLayerService.js [hunk 22]
```diff
@@ -0,0 +1,594 @@
...
+          encodeOverlay: getEncodeOverlay(gaUrlUtils),
+          encodeGraticule: encodeGraticule
+      };
+    };
+  };
+
+  function encodeGraticule(dpi) {
+      return {
+          'baseURL': 'https://wms.geo.admin.ch/',
+          'opacity': 1,
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

File: src/components/print/PrintLayerService.js [hunk 23]
```diff
@@ -0,0 +1,594 @@
...
+          }
+      };
+  };
+
+  function getEncodeOverlay(gaUrlUtils) {
+    return encodeOverlay;
+
+    function encodeOverlay(overlay, resolution, options) {
+
+        var encOverlayLayer;
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

File: src/components/print/PrintLayerService.js [hunk 24]
```diff
@@ -0,0 +1,594 @@
...
+        var maxResolution = layerConfig.maxResolution || Infinity;
+
+        if (resolution <= maxResolution &&
+            resolution >= minResolution) {
+          if (src instanceof ol.source.WMTS) {
+            var encodeWMTS = getEncodeWMTS(gaTime, gaMapUtils);
+            encLayer = encodeWMTS(layer, layerConfig);
+          } else if (src instanceof ol.source.ImageWMS ||
+              src instanceof ol.source.TileWMS) {
+                encLayer = encodeWMS(layer,
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

File: src/components/print/PrintLayerService.js [hunk 25]
```diff
@@ -0,0 +1,594 @@
...
+
+       angular.extend(enc, {
+         layer: source.getLayer(),
+         baseURL: baseUrl,
+         matrixIds: matrices
+       });
+     }
+
+     var multiPagesPrint = false;
+       if (config.timestamps) {
```
Comment (oterral) [inline src/components/print/PrintLayerService.js]: indent

**Comment by oterral** [comment]
Stille a lot of indent issue

**Comment by procrastinatio** [comment]
jenkins tests

**Comment by procrastinatio** [comment]
jenkins test

**Comment by gjn** [comment]
Thanks alot for this. I like the separation and testability.

**Comment by gjn** [comment]
Not all prints are working. E.g, [this can't be printed](https://mf-geoadmin3.int.bgdi.ch/?lang=de&topic=ech&bgLayer=ch.swisstopo.pixelkarte-grau&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bav.haltestellen-oev,ch.swisstopo.swisstlm3d-wanderwege,ch.swisstopo.fixpunkte-agnes,ch.astra.unfaelle-personenschaeden_getoetete,ch.bafu.hydroweb-messstationen_vorhersage&layers_visibility=true,false,false,false,false,true,false&layers_timestamp=20041231,,,,,,&X=198483.89&Y=600581.42&zoom=7&time=2004)

It think it involved time specific stuff.

I'll revert this PR.

**Comment by procrastinatio** [comment]
And where is this branch?!??

**Comment by gjn** [comment]
https://github.com/geoadmin/mf-geoadmin3/commits/mom_print_service
